File: Examples/SquareEtch/SquareEtch.cpp
#include <iostream>

#include <lsAdvect.hpp>
#include <lsBooleanOperation.hpp>
#include <lsDomain.hpp>
#include <lsExpand.hpp>
#include <lsMakeGeometry.hpp>
#include <lsPrune.hpp>
#include <lsToMesh.hpp>
#include <lsToSurfaceMesh.hpp>
#include <lsVTKWriter.hpp>

/**
  3D Example showing how to use the library for topography
  simulation, by creating a square trench geometry. A patch
  of exposed substrate is etched directionally.
  \example SquareEtch.cpp
*/

// Numerical velocity field.
// Advection scheme will take care of numerical
// artefacts itself.
class velocityField : public lsVelocityField<double> {
public:
  double getScalarVelocity(const std::array<double, 3> & /*coordinate*/,
                           int material,
                           const std::array<double, 3> &normalVector) {
    // if the surface of material 1 is facing upwards, etch it anisotropically
    if (material == 1 && normalVector[1] > 0.) {
      return -std::abs(normalVector[1]);
    } else
      return 0.;
  }
};

// Same velocity field, but analytical
// If the dissipation alphas can be derived,
// this will produce better results than numerical
// approximations. lsLocalLaxFriedrichsAnalytical has
// to be used for advection.
class analyticalField : public lsVelocityField<double> {
  const double velocity = -1;

public:
  double getScalarVelocity(const std::array<double, 3> & /*coordinate*/,
                           int material,
                           const std::array<double, 3> &normalVector) {
    if (material != 1)
      return 0.;

    return velocity * std::abs(normalVector[1]);
  }

  double getDissipationAlpha(int direction, int material,
                             const std::array<double, 3> &centralDifferences) {
    if (material != 1)
      return 0;

    double gradient = 0.;
    for (unsigned i = 0; i < 3; ++i) {
      gradient += centralDifferences[i] * centralDifferences[i];
    }
    gradient = std::sqrt(gradient);

    // alpha for different directions
    if (direction == 0) {
      return 0;
    } else if (direction == 1) {
      return std::abs(velocity);
    } else {
      return 0;
    }
  }
};

int main() {

  constexpr int D = 2;
  omp_set_num_threads(1);

  // Change this to use the analytical velocity field
  const bool useAnalyticalVelocity = false;

  double extent = 30;
  double gridDelta = 0.47;

  double bounds[2 * D] = {-extent, extent, -extent,
                          extent}; //, -extent, extent};
  lsDomain<double, D>::BoundaryType boundaryCons[D];
  for (unsigned i = 0; i < D - 1; ++i)
    boundaryCons[i] = lsDomain<double, D>::BoundaryType::REFLECTIVE_BOUNDARY;
  boundaryCons[D - 1] = lsDomain<double, D>::BoundaryType::INFINITE_BOUNDARY;

  auto substrate =
      lsSmartPointer<lsDomain<double, D>>::New(bounds, boundaryCons, gridDelta);

  double origin[3] = {0., 0., 0.};
  double planeNormal[3] = {0., D == 2, D == 3};
  {
    auto plane = lsSmartPointer<lsPlane<double, D>>::New(origin, planeNormal);
    lsMakeGeometry<double, D>(substrate, plane).apply();
  }

  double trenchBottom = -2.;
  {
    auto trench = lsSmartPointer<lsDomain<double, D>>::New(bounds, boundaryCons,
                                                           gridDelta);
    // trench bottom is the initial bottom of the trench
    double minCorner[D] = {-extent / 1.5, trenchBottom};
    double maxCorner[D] = {extent / 1.5, 1.};
    auto box = lsSmartPointer<lsBox<double, D>>::New(minCorner, maxCorner);
    lsMakeGeometry<double, D>(trench, box).apply();

    // Create trench geometry
    lsBooleanOperation<double, D>(substrate, trench,
                                  lsBooleanOperationEnum::RELATIVE_COMPLEMENT)
        .apply();
  }

  // in order only to etch the bottom of the trench, we need a mask layer
  auto mask =
      lsSmartPointer<lsDomain<double, D>>::New(bounds, boundaryCons, gridDelta);
  // make downward facing plane to remove bottom of trench for the mask
  // layer
  // add small offset so bottom of trench is definetly gone
  origin[D - 1] = trenchBottom + 1e-9;
  planeNormal[D - 1] = -1.;
  lsMakeGeometry<double, D>(
      mask, lsSmartPointer<lsPlane<double, D>>::New(origin, planeNormal))
      .apply();
  lsBooleanOperation<double, D>(mask, substrate,
                                lsBooleanOperationEnum::INTERSECT)
      .apply();

  std::string fileName;
  if (useAnalyticalVelocity)
    fileName = "analytical-";
  else
    fileName = "numerical-";
  {
    std::cout << "Extracting..." << std::endl;
    // output substrate layer (which wraps around mask layer)
    // wrapping is necessary for stable advection
    auto mesh = lsSmartPointer<lsMesh>::New();
    lsToSurfaceMesh<double, D>(substrate, mesh).apply();
    lsVTKWriter(mesh, fileName + "0.vtk").apply();

    // output mask layer
    lsToSurfaceMesh<double, D>(mask, mesh).apply();
    lsVTKWriter(mesh, "mask.vtk").apply();
  }

  // START ADVECTION
  auto velocities = lsSmartPointer<velocityField>::New();
  lsSmartPointer<analyticalField> analyticalVelocities;

  std::cout << "Advecting" << std::endl;
  lsAdvect<double, D> advectionKernel;

  // the level set to be advected has to be inserted last
  // the other is used as the mask layer for etching
  advectionKernel.insertNextLevelSet(mask);
  advectionKernel.insertNextLevelSet(substrate);
  advectionKernel.setSaveAdvectionVelocities(true);

  if (useAnalyticalVelocity) {
    advectionKernel.setVelocityField(analyticalVelocities);
    // Analytical velocity fields and dissipation coefficients
    // can only be used with this integration scheme
    advectionKernel.setIntegrationScheme(
        lsIntegrationSchemeEnum::LOCAL_LAX_FRIEDRICHS_ANALYTICAL_1ST_ORDER);
  } else {
    // for numerical velocities, just use the default
    // integration scheme, which is not accurate for certain
    // velocity functions but very fast
    advectionKernel.setVelocityField(velocities);

    // For coordinate independent velocity functions
    // this numerical scheme is superior though.
    // However, it is slower.
    // advectionKernel.setIntegrationScheme(
    //     lsIntegrationSchemeEnum::STENCIL_LOCAL_LAX_FRIEDRICHS_1ST_ORDER);
  }

  // advect the level set until 50s have passed
  double finalTime = 50;
  unsigned counter = 1;
  for (double time = 0.; time < finalTime;
       time += advectionKernel.getAdvectedTime()) {
    advectionKernel.apply();
    std::cout << "Advection step: " << counter
              << ", time: " << advectionKernel.getAdvectedTime() << std::endl;

    auto mesh = lsSmartPointer<lsMesh>::New();
    lsToSurfaceMesh<double, D>(substrate, mesh).apply();
    lsVTKWriter(mesh, fileName + std::to_string(counter) + ".vtk").apply();
    ++counter;
  }
  std::cout << std::endl;
  std::cout << "Number of Advection steps taken: " << counter << std::endl;

  auto mesh = lsSmartPointer<lsMesh>::New();
  lsToSurfaceMesh<double, D>(substrate, mesh).apply();
  lsVTKWriter(mesh, "final.vtk").apply();

  return 0;
}

File: examples/SquareEtch/SquareEtch.py
import viennals as vls
import argparse

vls.Logger.setLogLevel(vls.LogLevel.INFO)


# 1. Define the Velocity Field
# We inherit from viennals.VelocityField to define custom logic in Python
class EtchingField(vls.VelocityField):
    def getScalarVelocity(self, coordinate, material, normalVector, pointId):
        if material == 2:
            return -0.1
        if material == 1:
            return -1.0
        return 0.0

    def getVectorVelocity(self, coordinate, material, normalVector, pointId):
        return [0.0] * len(coordinate)


# 1. Define the Velocity Field
# We inherit from viennals.VelocityField to define custom logic in Python
class DepositionField(vls.VelocityField):
    def getScalarVelocity(self, coordinate, material, normalVector, pointId):
        return 0.75

    def getVectorVelocity(self, coordinate, material, normalVector, pointId):
        return [0.0] * len(coordinate)


def main():
    # 1. Parse Arguments
    parser = argparse.ArgumentParser(
        description="Run Square Etch simulation in 2D or 3D."
    )
    parser.add_argument(
        "-D",
        "--dim",
        type=int,
        default=2,
        choices=[2, 3],
        help="Dimension of the simulation (2 or 3). Default is 2.",
    )
    args = parser.parse_args()

    DIMENSION = args.dim
    vls.setDimension(DIMENSION)
    vls.setNumThreads(8)

    extent = 30.0
    gridDelta = 0.47

    # Define bounds and boundary conditions
    trenchBottom = -2.0
    bounds = [-extent, extent, -extent, extent]
    origin = [0.0, 0.0]
    if DIMENSION == 3:
        bounds.append(-extent)
        bounds.append(extent)
        origin.append(0.0)
    boundaryCons = []
    for i in range(DIMENSION - 1):
        boundaryCons.append(vls.BoundaryConditionEnum.REFLECTIVE_BOUNDARY)
    boundaryCons.append(vls.BoundaryConditionEnum.INFINITE_BOUNDARY)

    planeNormal = list(origin)
    downNormal = list(origin)
    planeNormal[DIMENSION - 1] = 1.0
    downNormal[DIMENSION - 1] = -1.0

    # Create the Substrate Domain
    substrate = vls.Domain(bounds, boundaryCons, gridDelta)

    # Create initial flat geometry (Plane at y/z=0)
    plane = vls.Plane(origin, planeNormal)
    vls.MakeGeometry(substrate, plane).apply()

    # --------------------------------------
    # 3. Trench Geometry
    # --------------------------------------
    trench = vls.Domain(bounds, boundaryCons, gridDelta)

    # Define Box Corners based on dimension
    minCorner = list(origin)
    maxCorner = list(origin)
    originMask = list(origin)
    for i in range(DIMENSION - 1):
        minCorner[i] = -extent / 1.5
        maxCorner[i] = extent / 1.5
    minCorner[DIMENSION - 1] = trenchBottom
    maxCorner[DIMENSION - 1] = 1.0
    originMask[DIMENSION - 1] = trenchBottom + 1e-9

    box = vls.Box(minCorner, maxCorner)
    vls.MakeGeometry(trench, box).apply()

    # Subtract trench from substrate (Relative Complement)
    vls.BooleanOperation(
        substrate, trench, vls.BooleanOperationEnum.RELATIVE_COMPLEMENT
    ).apply()

    # 4. Create the Mask Layer
    # The mask prevents etching at the bottom of the trench in specific configurations,
    # or acts as a hard stop/masking material.
    mask = vls.Domain(bounds, boundaryCons, gridDelta)

    vls.MakeGeometry(mask, vls.Plane(originMask, downNormal)).apply()

    # Intersect with substrate geometry
    vls.BooleanOperation(mask, substrate, vls.BooleanOperationEnum.INTERSECT).apply()

    # 5. Export Initial State
    print(f"Running in {DIMENSION}D.")
    print("Extracting initial meshes...")
    mesh = vls.Mesh()

    # Save substrate
    vls.ToSurfaceMesh(substrate, mesh).apply()
    vls.VTKWriter(mesh, f"substrate-{DIMENSION}D.vtp").apply()

    # Save mask
    vls.ToSurfaceMesh(mask, mesh).apply()
    vls.VTKWriter(mesh, f"mask-{DIMENSION}D.vtp").apply()

    print("Creating new layer...")
    polymer = vls.Domain(substrate)

    # 6. Setup Advection
    deposition = DepositionField()
    etching = EtchingField()

    print("Advecting")
    advectionKernel = vls.Advect()

    # the level set to be advected has to be inserted last
    # the other could be taken as a mask layer for advection
    advectionKernel.insertNextLevelSet(mask)
    advectionKernel.insertNextLevelSet(substrate)
    advectionKernel.insertNextLevelSet(polymer)

    advectionKernel.setSaveAdvectionVelocities(True)
    # advectionKernel.setVelocityField(etching)

    advectionKernel.setVelocityField(deposition)
    advectionKernel.apply()

    vls.ToSurfaceMesh(polymer, mesh).apply()
    vls.VTKWriter(mesh, f"newLayer-{DIMENSION}D.vtp").apply()

    advectionKernel.setSaveAdvectionVelocities(True)
    advectionKernel.setVelocityField(etching)

    # Use default spatial discretization scheme (Lax Friedrichs 1st order) as in the C++ else branch
    # advectionKernel.setSpatialScheme(viennals.SpatialSchemeEnum.LAX_FRIEDRICHS_1ST_ORDER)

    # 7. Time Loop
    finalTime = 10.0
    counter = 1
    currentTime = 0.0

    # The advect kernel calculates stable time steps automatically.
    # We call apply() repeatedly until we reach finalTime.

    # Note: In the C++ example, the loop structure is:
    # for (double time = 0.; time < finalTime; time += advectionKernel.getAdvectedTime())
    # This implies one step is taken, then time is updated.

    # Save initial state
    vls.ToSurfaceMesh(polymer, mesh).apply()
    vls.VTKWriter(mesh, f"numerical-{DIMENSION}D-0.vtp").apply()

    while currentTime < finalTime:
        advectionKernel.apply()

        stepTime = advectionKernel.getAdvectedTime()
        currentTime += stepTime

        print(
            f"Advection step: {counter}, time: {stepTime:.4f} (Total: {currentTime:.4f})"
        )

        # Export result
        vls.ToSurfaceMesh(polymer, mesh).apply()
        vls.VTKWriter(mesh, f"numerical-{DIMENSION}D-{counter}.vtp").apply()

        counter += 1

    print(f"\nNumber of Advection steps taken: {counter}")

    # Final export
    vls.ToSurfaceMesh(polymer, mesh).apply()
    vls.VTKWriter(mesh, f"final-{DIMENSION}D.vtp").apply()


if __name__ == "__main__":
    main()
